File: golang/client/socket_worker.go
package client

import (
	"errors"
	"net"
	"strconv"
	"strings"
	"time"

	"github.com/TorkhanNetwork/Networking/golang/data_encryption"
	"github.com/google/uuid"
	"github.com/kataras/golog"
)

const INPUT_BUFFER_SIZE = 65536

type SocketWorker struct {
	Id                                           int
	client                                       Client
	targetIp, commandPrefix, requestSeparator    string
	targetPort                                   int
	connection                                   net.TCPConn
	connected, authenticated, connectionProtocol bool
	keyGenerator                                 data_encryption.KeyGenerator
	quitChan                                     chan bool
}

func NewSocketWorker(id int, client Client, targetIp string, targetPort int) SocketWorker {
	return SocketWorker{
		Id:           id,
		client:       client,
		targetIp:     targetIp,
		targetPort:   targetPort,
		keyGenerator: data_encryption.NewGenerator(),
	}
}

func (socketWorker SocketWorker) GetName() string {
	return "SocketWorker #" + strconv.Itoa(socketWorker.Id)
}

func (socketWorker *SocketWorker) StartWorker() {
	if err := socketWorker.ConnectSocket(); err != nil {
		golog.Error(socketWorker.GetName()+" - Unbale to start worker\n", err)
		return
	}
	socketWorker.quitChan = make(chan bool)
	go socketWorker.handleServerSocket()
}

func (socketWorker *SocketWorker) StopWorker() {
	if socketWorker.connected {
		socketWorker.SendCommand("disconnect")
		socketWorker.quitChan <- true
	}
}

func (socketWorker *SocketWorker) ConnectSocket() error {
	if socketWorker.connected {
		return errors.New(socketWorker.GetName() + " - Socket is already connected")
	}
	tcpAddr, err := net.ResolveTCPAddr("tcp", socketWorker.targetIp+":"+strconv.Itoa(socketWorker.targetPort))
	if err != nil {
		return err
	}
	connection, err := net.DialTCP("tcp", nil, tcpAddr)
	if err != nil {
		return err
	}
	if connection == nil {
		return errors.New(socketWorker.GetName() + " - Unable to connect the socket")
	}
	connection.SetDeadline(time.Time{})
	socketWorker.connection = *connection
	socketWorker.connected = true
	return nil
}

func (socketWorker *SocketWorker) DisconnectSocket() error {
	if !socketWorker.connected {
		return errors.New(socketWorker.GetName() + " - Socket isn't connected")
	}
	err := socketWorker.connection.Close()
	if err != nil {
		return err
	}
	socketWorker.StopWorker()
	socketWorker.connected = false
	socketWorker.client.RemoveSocketWorker(*socketWorker)
	return nil
}

func (socketWorker *SocketWorker) handleServerSocket() {
	buf := make([]byte, INPUT_BUFFER_SIZE)
	socketWorker.startConnectionProtocol()
	for {
		mLen, err := socketWorker.connection.Read(buf)
		if err != nil {
			break
		}
		select {
		case <-socketWorker.quitChan:
			socketWorker.DisconnectSocket()
		default:
			for _, line := range strings.Split(string(buf[:mLen]), "\n") {
				socketWorker.onLineRead(line)
			}
		}
	}

}

func (socketWorker *SocketWorker) onLineRead(line string) {
	line = strings.TrimSpace(line)
	if line != "" {
		if decryptedData, err := data_encryption.Decrypt(line, socketWorker.keyGenerator); err == nil {
			socketWorker.onDataReceived([]string{decryptedData, line}, true)
		} else {
			socketWorker.onDataReceived([]string{line}, false)
		}
	}

}

func (socketWorker *SocketWorker) onDataReceived(data []string, encrypted bool) {
	var msgUUID uuid.UUID
	if strings.HasPrefix(data[0], "response:") {
		msgUUID, err := uuid.Parse(data[0][9:45])
		if err != nil {
			golog.Error(socketWorker.GetName()+" - Unable to parse response UUID\n", err)
			return
		}
		data[0] = data[0][45:]
		golog.Debug(socketWorker.GetName() + " - Response received for UUID " + msgUUID.String())
	} else {
		var err error
		msgUUID, err = uuid.Parse(data[0][:36])
		if err != nil {
			golog.Error(socketWorker.GetName()+" - Unable to parse response UUID\n", err)
			return
		}
		data[0] = data[0][36:]
	}

	if socketWorker.commandPrefix != "" && strings.HasPrefix(data[0], socketWorker.commandPrefix) {
		socketWorker.onCommandReceived(data[0], msgUUID, encrypted)
	} else {
		golog.Debug(socketWorker.GetName() + " - Data received (encrypted=" + strconv.FormatBool(encrypted) + ", uuid=" + msgUUID.String() + ") : " + data[0])
	}

	// TODO event

	if socketWorker.connectionProtocol {
		if !socketWorker.onConnectionProtocolDataReceived(data[0]) {
			golog.Error(socketWorker.GetName() + " - Unable to connect to the server")
			socketWorker.StopWorker()
		}
	}

}

func (socketWorker *SocketWorker) onCommandReceived(command string, msgUUID uuid.UUID, encrypted bool) {
	golog.Debug(socketWorker.GetName() + " - Command Received (encrypted=" + strconv.FormatBool(encrypted) + ", uuid=" + msgUUID.String() + ") : " + command)
	// TODO Event
}

func (socketWorker *SocketWorker) startConnectionProtocol() {
	socketWorker.connectionProtocol = true
	socketWorker.SendData("version:"+CLIENT_VERSION, uuid.NullUUID{}, false)
}

func (socketWorker *SocketWorker) onConnectionProtocolDataReceived(data string) bool {
	if strings.HasPrefix(data, "version:") {
		if version := strings.ToLower(data[8:]) == "true"; version {
			socketWorker.SendData("macAddress:"+socketWorker.client.MacAddress, uuid.NullUUID{}, false)
		} else {
			return false
		}
	} else if strings.HasPrefix(data, "publicKey:") {
		pub, err := ParseRsaPublicKeyFromPemStr(data[10:])
		if err != nil {
			golog.Error(socketWorker.GetName()+" - Failed to parse server public key\n", err)
			return false
		}
		socketWorker.keyGenerator.PublicKey = *pub
		socketWorker.keyGenerator.GenerateKeys(true, false)
		secretKey, err := data_encryption.EncryptSecretKey(socketWorker.keyGenerator)
		if err != nil {
			golog.Error("Unable to encrypt secret key\n", err)
			return false
		}
		socketWorker.SendData("secretKey:"+secretKey, uuid.NullUUID{}, false)
	} else if strings.HasPrefix(data, "separator:") {
		socketWorker.requestSeparator = data[9:]
	} else if strings.HasPrefix(data, "commandPrefix:") {
		socketWorker.commandPrefix = data[14:]
	} else if strings.HasPrefix(data, "connected:") {
		connected := strings.ToLower(data[10:]) == "true"
		socketWorker.authenticated = connected
		socketWorker.connectionProtocol = !connected
		// TODO event
		if connected {
			socketWorker.SendCommand("blou")
		}
		return connected
	} else {
		return false
	}
	return true
}

func (socketWorker *SocketWorker) SendData(data string, responseUUID uuid.NullUUID, encrypt bool) {
	rawData := data
	// TODO response
	data = uuid.New().String() + data
	if encrypt {
		var err error
		data, err = data_encryption.Encrypt(data, socketWorker.keyGenerator)
		if err != nil {
			golog.Error(socketWorker.GetName()+" - Unable to encrypt data\n", err)
			return
		}
	}
	_, err := socketWorker.connection.Write([]byte(data + "\n"))
	if err != nil {
		golog.Error(socketWorker.GetName()+" - Unable to send data\n", err)
		return
	}
	golog.Debug(socketWorker.GetName() + " - Data sent (encrypted=" + strconv.FormatBool(encrypt) + ", uuid=not supported yet) : " + rawData)
}

func (socketWorker *SocketWorker) SendCommand(command string, args ...string) {
	socketWorker.SendData(socketWorker.commandPrefix+command+socketWorker.requestSeparator+strings.Join(args, socketWorker.requestSeparator), uuid.NullUUID{}, true)
}

File: python/client/SocketWorker.py
import socket as socketlib
import threading
import uuid

from libs.response_system import IResponseWorker, Response
from SocketStreamReader import SocketStreamReader
from libs.data_encryption import KeyGenerator, EncryptionManager


class SocketWorker(IResponseWorker.IResponseWorker):
    def __init__(self, client, ip, port):
        IResponseWorker.IResponseWorker.__init__(self)
        self.client = client
        self.ip = ip
        self.port = port
        self.socket = None
        self.connected = False
        self.authenticated = False
        self.reader = None
        self.keyGenerator = KeyGenerator.KeyGenerator()
        self.commandPrefix = None
        self.requestSeparator = None
        self.connectionProtocol = False

    def startWorker(self):
        self.__connectSocket()
        self.start()
        self.name = "Socket Worker #" + str(self.native_id)

    def stopWorker(self):
        if self.connected:
            self.sendData("disconnect", encrypt=False)
            self.interrupt()

    def __connectSocket(self):
        self.socket = socketlib.socket()
        self.socket.connect((self.ip, self.port))
        self.reader = SocketStreamReader(self.socket)
        self.connected = True
        self.client.logger.log("Socket is connected")
        # socket.send(bytes("env:" + self.client.getPythonVersion() + "\n", 'utf-8'))
        # socket.send(b"version:1.1.0\n")
        # socket.close()

    def __disconnectSocket(self):
        if self.socket is not None:
            self.socket.close()
            self.connected = False
            self.interrupt()
            self.socket = None

    def run(self) -> None:
        self.__handleServerSocket()

    def __handleServerSocket(self):
        self.__startConnectionProtocol()
        t = threading.currentThread()
        while getattr(t, "do_run", True):
            line = self.reader.readLine().replace("\n", "").replace("\r", "")
            if line is not None:
                self.__onLineRead(line)
            else:
                break
        self.__disconnectSocket()

    def __onLineRead(self, line: str):
        decryptedData = None
        if self.keyGenerator.secretKey is not None:
            decryptedData = EncryptionManager.decrypt(
                line, self.keyGenerator)
        if decryptedData is not None:
            self.__onDataReceived([decryptedData, line], True)
        else:
            self.__onDataReceived([line], False)

    def __onDataReceived(self, data: list, encrypted: bool):
        if data[0].startswith("response:"):
            msgUUID = uuid.UUID(data[0][9:45])
            data[0] = data[0][45:]
            self.client.logger.log(
                self.name + " - Response received for UUID " + str(msgUUID))
            self.responseManager.onResponseReceived(msgUUID, data[0])
        else:
            msgUUID = uuid.UUID(data[0][:36])
            data[0] = data[0][36:]

        if self.commandPrefix is not None and str(data[0]).startswith(self.commandPrefix):
            self.__onCommandReceived(data[0], msgUUID, encrypted)
        else:
            self.client.logger.log(self.name + " - Data Received (encrypted=" + str(encrypted) + ", uuid=" + str(msgUUID) + ") : "
                                   + data[0])

            # todo event system

            if self.connectionProtocol and not (self.__onConnectionProtocolDataReceived(data[0])):
                self.client.logger.log(
                    self.name + " - Unable to connect to the server")
                self.stopWorker()

    def __onCommandReceived(self, command: str, msgUUID: uuid.UUID, encrypted: bool):
        self.client.logger.log(self.name + " - Command Received (encrypted=" + str(
            encrypted) + ", uuid=" + str(msgUUID) + ") : "
            + command)

        # todo event system

    def __startConnectionProtocol(self):
        self.connectionProtocol = True
        self.sendData("version:" + self.client.getVersion(), encrypt=False)

    def __onConnectionProtocolDataReceived(self, data: str) -> bool:
        if data.startswith("version:"):
            version = data[8:] == "true"
            if not version:
                return False
            self.sendData("macAddress:" +
                          self.client.getMacAddress(), encrypt=False)
        elif data.startswith("publicKey:"):
            self.keyGenerator.publicKey = EncryptionManager.decryptPublicKey(
                data[10:])
            self.keyGenerator.generateKeys(True, False)
            self.sendData(
                "secretKey:" + EncryptionManager.encryptSecretKey(self.keyGenerator), encrypt=False)
        elif data.startswith("separator:"):
            self.requestSeparator = data[10:]
        elif data.startswith("commandPrefix:"):
            self.commandPrefix = data[14:]
        elif data.startswith("connected:"):
            connected = data[10:] == "true"
            self.authenticated = connected
            self.connectionProtocol = not connected
            if connected:
                # todo event system
                pass
            return connected
        else:
            return False
        return True

    def sendData(self, data: str, msgUUID: uuid.UUID = None, encrypt: bool = True):
        rawData = str(data)
        response = Response.Response(self, msgUUID)
        if msgUUID is not None:
            data = "response:" + data
        data = str(response.msgUUID) + data
        if encrypt:
            data = EncryptionManager.encrypt(
                data, self.keyGenerator)
        self.socket.send(bytes(data + "\n", 'utf-8'))
        self.client.logger.log(self.name + " - Data sent (encrypted=" + str(encrypt) + ", uuid=" + str(response.msgUUID) + ") : "
                               + rawData)
        return response

    def sendCommand(self, command: str, args: tuple = None):
        return self.sendData(self.commandPrefix + command + self.requestSeparator + (self.requestSeparator.join(args) if args is not None else ""))

    def interrupt(self):
        setattr(threading.currentThread(), 'do_run', False)
